### PR TITLE
[dns] handle instance names with dots

### PR DESCRIPTION
**Pull request by simonlingoogle**
This commit adds a new method `SplitFullDnsName` which handles instance names with dots correctly. 

- [x] Add a unit test.

### CODE REVIEW

File: src/common/dns_utils.cpp [hunk 1]
```diff
@@ -27,88 +27,114 @@
...
 #include "common/dns_utils.hpp"
+
+#include <assert.h>
+
 #include "common/code_utils.hpp"
 
-DnsNameType GetDnsNameType(const std::string &aFullName)
+static bool NameEndsWithDot(const std::string &aName)
+{
+    return !aName.empty() && aName[aName.length() - 1] == '.';
```
Comment (wgtdkp) [inline src/common/dns_utils.cpp:37]:
```suggestion
    return !aName.empty() && aName.back() == '.';
```

**Comment by simonlingoogle** [comment]
Unit test added @wgtdkp

**Review by wgtdkp — APPROVED** [review]
LGTM :+1:

**Review by abtink — APPROVED** [review]
LGTM. 👍

File: src/common/dns_utils.hpp [hunk 2]
```diff
@@ -37,26 +37,54 @@
...
-enum DnsNameType
+struct DnsNameInfo
 {
-    kDnsNameTypeUnknown,
-    kDnsNameTypeService,
-    kDnsNameTypeInstance,
-    kDnsNameTypeHost,
+    std::string mInstanceName; ///< Instance name, or empty if the DNS name is not a service instance.
+    std::string mServiceName;  ///< Service name, or empty if the DNS name is not a service or service instance.
+    std::string mHostName;     ///< Host name.
```
Comment (jwhui) [inline src/common/dns_utils.hpp:49]: nit: can this be empty?

**Review by jwhui — COMMENTED** [review]
Looks great overall! One minor nit.

File: src/common/dns_utils.hpp (hunk 2 above)
Comment (simonlingoogle) [reply src/common/dns_utils.hpp:49]: revised

**Review by librasungirl — APPROVED** [review]

**Review by jwhui — APPROVED** [review]

**Pull request merged**